mesonbuild/options.py

```diff
@@ -940,6 +925,19 @@ def add_project_option(self, key: T.Union[OptionKey, str], valobj: AnyOptionType
         assert key.subproject is not None
         if key in self.options:
             raise MesonException(f'Internal error: tried to add a project option {key} that already exists.')
+        if valobj.yielding and key.subproject:
+            parent_key = key.as_root()
+            try:
+                parent_option = self.options[parent_key]
+                # If parent object has different type, do not yield.
+                # This should probably be an error.
```

Comment:
This comment is actually incorrect, because you can easily have two subprojects with yielding options of different types. The people who consume subprojects cannot be expected to patch that out, especially because subprojects can be instantiated recursively. At best we can warn about it, but a non-fatal warning.

Reply:
Makes sense, though yielding options should be limited to a few well known ones and are almost always boolean so the fallout should be minimal. 

I just moved the comment and didn't want to open the can of worms though.